tensorboard/plugins/hparams/hparams_plugin.py

```diff
@@ -90,6 +102,107 @@ def is_active(self):
+            self._context, request_proto
+        ).run()
+        experiment = get_experiment.Handler(self._context).run()
+        try:
+            body, mime_type = self.download_data_impl(
+                experiment, session_groups, response_format
+            )
+        except ValueError as e:
+            return http_util.Respond(
+                request=request,
+                content=str(e),
+                content_type="text/plain",
+                code=400,
+            )
+        return http_util.Respond(request, body, mime_type)
+
+    def download_data_impl(self, experiment, session_groups, response_format):
+        """Provides a response for downloading data for an HParams table.
+
+        Args:
+          experiment: Experiment proto.
+          session_groups: ListSessionGroupsResponse proto.
+          response_format: A string in the OutputFormat enum.
+
+        Raises:
+          ValueError: If the scalars plugin is not registered.
+
+        Returns:
+          2 entities:
+            - A JSON object response body.
+            - A mime type (string) for the response.
+        """
+        header = []
+        for hparam_info in experiment.hparam_infos:
+            header.append(hparam_info.display_name or hparam_info.name)
+
+        for metric_info in experiment.metric_infos:
+            header.append(metric_info.display_name or metric_info.name.tag)
+
+        rows = []
+
+        def _get_value(value):
+            if value.HasField("number_value"):
+                return value.number_value
+            if value.HasField("string_value"):
+                return value.string_value
+            if value.HasField("bool_value"):
+                return value.bool_value
+            return None
+
+        for group in session_groups.session_groups:
+            row = []
+            for hparam_info in experiment.hparam_infos:
+                row.append(_get_value(group.hparams[hparam_info.name]))
+            for metric_value in group.metric_values:
+                row.append(metric_value.value)
+            rows.append(row)
+
+        if response_format == OutputFormat.JSON:
+            mime_type = "application/json"
+            body = dict(header=header, rows=rows)
+        elif response_format == OutputFormat.LATEX:
+
+            def latex_format(value):
+                if isinstance(value, int):
+                    return str(value)
+                elif isinstance(value, float):
+                    return "%.3g" % value
+                return value.replace("_", "\\_").replace("%", "\\%")
+
+            mime_type = "application/x-latex"
+            top_part = "\\begin{table}[]\n\\begin{tabular}{%s}\n" % (
+                "l" * len(header)
+            )
+            header_part = (
+                " & ".join(map(latex_format, header)) + " \\\\ \\hline\n"
+            )
+            middle_part = "".join(
+                [" & ".join(map(latex_format, row)) + " \\\\\n" for row in rows]
```

Comment:
Can drop the list comprehension to use a generator comprehension
directly, which avoids allocating the intermediate list:

```python
            middle_part = "".join(
                " & ".join(map(latex_format, row)) + " \\\\\n" for row in rows
            )
```